### ISSUE TITLE
 Proposal to always build the measured boot image feature

### ISSUE THREAD

**Issue by wainersm** [kata-containers/kata-containers#7415]
>Note: this issue is based on a discussion that @fidencio, @stevenhorsman and I had about the merge of CCv0 and main branches. This is basically a re-write of @stevenhorsman 's comment on https://github.com/kata-containers/kata-containers/pull/7231#issuecomment-1643944633 with a couple of extra information.

**Problem statement**

The [measured boot image/rootfs](https://github.com/confidential-containers/documentation/issues/40) feature was proposed by @arronwy on the context of Confidential Containers (CoCo) aiming to provide an efficient (performance wise) way to check the integrity of the guest image/rootfs.

In a nutshell it works like:

 * an [initramfs that contains the tooling](https://github.com/kata-containers/kata-containers/tree/main/tools/packaging/static-build/initramfs) (e.g. `veritysetup`)  to verify guest image on boot time should be built
 * the initramfs is built into the guest kernel
 * the initramfs's init scripts mount the guest rootfs and perform checks
 * the measured boot is activated only and only if the `rootfs_verity.scheme` and `cc_rootfs_verity.hash` parameters are passed to the kernel (down to the initramfs's init) via Kata's `configuration.toml` (`kernel_params` option). 

That feature is currently enabled on build time by exporting the `MEASURED_BOOT=yes` variable. It has been enabled by default on "vanilla" and TDX builds of Kata Containers on the CCv0 branch. The SEV/SNP builds doesn't build on that because they have already another TEE-specific approach for boot measurement, thus, SEV/SNP build with `MEASURED_BOOT=no`. On the `main` branch, despite the measured boot image code already be there, currently it isn't tested by any CI job.

All that said, our concern is about how to handle that feature on the `main` branch. After some chatting, we agreed that the best approach is what is described on the next section.

**Which feature do you think can be improved?**

The measured boot rootfs is hardware-agnostic so you don't need an TEE (Trusted Execution Environment) like Intel TDX or AMD SEV to enable it. We believe that it can be useful outside of CoCo uses case, for example, an infra-owner can enable the measured boot image to check the kata's guest image is not tampered.

Our proposal comprises of two improvements:

1.  Adapt the build scripts to **always** build the initramfs into the kernel, even for the "vanilla" kernels
2.  Provides a mechanism to allow the user to easily enable the measured boot on a Kata Containers installation

**How can it be improved?**

Here I will be quoting @stevenhorsman (https://github.com/kata-containers/kata-containers/pull/7231#issuecomment-1643944633):

\'''
- As it is very difficult to calculate the root hash after the event, we can follow the current pattern of calculating it and pulling it into the runtime, but inject it to the config tomls but commented out:
  - We could have a new parameter that results in the kernel_params getting set like `debug_console_enabled=true` adds `agent.debug_console agent.debug_console_vport=1026` to the kernel_params and then a use would just have to uncomment that option to use it
  - Alternatively (and a bit cheaper, but not as nicer user experience), we could just add something like:
  ```
  # Add the following line to the kernel_params to enable integrity checking of the guest image 
  # rootfs_verity.scheme=dm-verity cc_rootfs_verity.hash=${root_hash}
  ```
  in the line above the kernel_params in the config

\'''
**Additional Information**

In order to take a decision whether or not always build with support, we should consider the following :

* the size of the kernel file will increase
* the kernel boot initialization will change as initramfs's init script will always run and switch to guest's rootfs `/sbin/init` eventually. Any performance concern?

## Test updates

We have a couple of tests for this feature in `CCV0` for both [kubernetes](https://github.com/kata-containers/tests/blob/087f2d01c2d4125927b27a60310f1d9115d1b96e/integration/kubernetes/confidential/agent_image.bats#L110-L125) and [containerd/crictl](https://github.com/kata-containers/tests/blob/087f2d01c2d4125927b27a60310f1d9115d1b96e/integration/containerd/confidential/agent_image.bats#L15-L32), but they are fairly tightly coupled with the other CC tests, so need refactoring and we'll want to move them into the kata-containers repo to be run with GHA
- We have an open-question of what platforms do we want to support for these tests? s390x, TDX etc?
- More comments and discussion in https://github.com/kata-containers/kata-containers/issues/7245, though it maybe outdated

## Acceptance Criteria

For each supported platform:

_**Scenario: Check correct hash works**_
**Given** I have a version of kata installed that has a kernel with the initramfs built and config with `rootfs_verity.scheme=dm-verity rootfs_verity.hash=<correct hash of rootfs>` set in the `kernel_params`
**When** I try and create a container a basic pod
**Then** The pod is successfully created

_**Scenario: Check incorrect hash fails**_
**Given** I have a version of kata installed that has a kernel with the initramfs built and config with `rootfs_verity.scheme=dm-verity rootfs_verity.hash=<incorrect hash of rootfs>` set in the `kernel_params`
**When** I try and create a container a basic pod 
**Then** The pod is doesn't run
  **And** Ideally we'd get a helpful message to indicate why

**Comment by ChengyuZhu6** [comment]
Maybe we can follow the current work to enable the feature, installation if necessary by exporting `MEASURED_BOOT` variable?

**Comment by stevenhorsman** [comment]
I guess we should add the tests (covered by https://github.com/kata-containers/kata-containers/issues/7245 but maybe out of date) as part of the PR that delivers this.

**Comment by wainersm** [comment]
Hi @stevenhorsman ,

I was looking at test case:

```
Scenario: Check incorrect hash fails
Given I have a version of kata installed that has a kernel with the initramfs built and config with rootfs_verity.scheme=dm-verity rootfs_verity.hash=<incorrect hash of rootfs> set in the kernel_params
When I try and create a container a basic pod
Then The pod is doesn't run
And Ideally we'd get a helpful message to indicate why
```

The current test case on CCv0 (`Test cannnot launch pod with measured boot enabled and rootfs modified`) actually modifies the guest image, i.e., it is a more realistic scenario than passing an incorrect hash. However modify the file on remote worker node will be.... you know. There is this k8s vs crictl tests question still open too.

I'm okay to follow the test case you proposed., and I will be working on that asap.  Just wanted to bring up that I will be not migrating the `Test cannnot launch pod with measured boot enabled and rootfs modified` to main though.

**Comment by stevenhorsman** [comment]
Thanks for the heads up. I do agree that the modified image is more realistic, but was trying to be pragmatic, maybe we could consider that for the crictl test, but that involves forking, so I'm not sure that's great either!

**Comment by stevenhorsman** [comment]
IIRC #7231 which closed this issue didn't cover the full scope of this - which was to have the measured rootfs value always calculated and added to the config, but commented out, so I'm re-opening this until that is done.

**Comment by Xynnn007** [comment]
Hi, has it been determined that the build of the guest image will definitely enable measured boot? 

CoCo Attestation should have reference values for the kernel cmdline items `cc_rootfs_verity.hash` and `rootfs_verity.scheme`. It would be a good place to publish the reference value file together with measured boot image to ghcr.io.

**Comment by wainersm** [comment]
Hi @Xynnn007 !

> Hi, has it been determined that the build of the guest image will definitely enable measured boot?
>

Currently only the TDX kernel  is built with measured rootfs. I got side-tracked by other stuffs and could not return to this topic and pursue the community to have an agreement whether to enable to all kernels or not. Well, maybe not all, but just the vanilla kernel because SEV, for example, has its own mechanism to measure measure boot and wouldn't be interested on the feature.

What's your opinion, @Xynnn007, BTW? 

> CoCo Attestation should have reference values for the kernel cmdline items `cc_rootfs_verity.hash` and `rootfs_verity.scheme`. It would be a good place to publish the reference value file together with measured boot image to ghcr.io.

I hadn't thought on the integration with CoCo Attestation, good point and it seems that publish on somewhere will be needed.

@danmihai1 has worked on measured rootfs as well (https://github.com/kata-containers/kata-containers/pull/7988). I feel that the measured rootfs use case(s) isn't well-defined yet, thus it is incomplete implemented. @arronwy @danmihai1 @Xynnn007 could we join forces here to better design/implement the feature?

Cc @fidencio @stevenhorsman

**Comment by Xynnn007** [comment]
Thanks @wainersm for the reply. 

> Currently only the TDX kernel is built with measured rootfs. I got side-tracked by other stuffs and could not return to this topic and pursue the community to have an agreement whether to enable to all kernels or not. Well, maybe not all, but just the vanilla kernel because SEV, for example, has its own mechanism to measure measure boot and wouldn't be interested on the feature.

> What's your opinion, @Xynnn007, BTW?

I am still not famlilar with the integrity checking for images on other platforms than TDX now. So sorry I cannot provide any useful suggestions now. : (


In order not to bring much information and context that annoy/interrupt you, I will then put an issue for the attestation things separately after mer-to-main work.

**Comment by stevenhorsman** [comment]
> @wainersm, can we consider this one fixed by #10466?

I don't think this PR's work goes as far as the proposal? Specifically there are two points it mentions including:

> 1. Adapt the build scripts to always build the initramfs into the kernel, even for the "vanilla" kernels
> 2. Provides a mechanism to allow the user to easily enable the measured boot on a Kata Containers installation

I think with #10466 you have enabled it just for the TDX and coco-dev runtimeclasses, not all the kernels?

There is an argument of whether measured rootfs is useful outside of CoCo and whether we want to have this feature on all kernels, but I'm not sure your PR (whilst great) solves it.

**Comment by ariel-adam** [comment]
@wainersm @stevenhorsman can we close this issue?

**Comment by stevenhorsman** [comment]
> [@wainersm](https://github.com/wainersm) [@stevenhorsman](https://github.com/stevenhorsman) can we close this issue?

We haven't implemented this, so why do you want to close it? Do you think it's not a useful requirement now?
